x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/DirectBlobContainerIndexInput.java

```diff
@@ -74,29 +85,53 @@ public DirectBlobContainerIndexInput(
         IndexInputStats stats,
         long sequentialReadSize
     ) {
-        this(name, blobContainer, fileInfo, context, stats, 0L, 0L, fileInfo.length(), sequentialReadSize);
+        this(name, blobContainer, fileInfo, bufferSize(context), stats, 0L, 0L, fileInfo.length(), sequentialReadSize);
         stats.incrementOpenCount();
     }
 
     private DirectBlobContainerIndexInput(
         String name,
         BlobContainer blobContainer,
         FileInfo fileInfo,
-        IOContext context,
+        int bufferSize,
         IndexInputStats stats,
         long position,
         long offset,
         long length,
         long sequentialReadSize
     ) {
-        super(logger, name, blobContainer, fileInfo, context, stats, offset, length); // TODO should use blob cache
+        super(name, bufferSize); // TODO should use blob cache
         this.position = position;
         assert sequentialReadSize >= 0;
         this.sequentialReadSize = sequentialReadSize;
+        this.blobContainer = Objects.requireNonNull(blobContainer);
+        this.fileInfo = Objects.requireNonNull(fileInfo);
+        assert fileInfo.metadata().hashEqualsContents() == false
+            : "this method should only be used with blobs that are NOT stored in metadata's hash field " + "(fileInfo: " + fileInfo + ')';
+        this.stats = Objects.requireNonNull(stats);
+        this.offset = offset;
+        this.length = length;
+        this.closed = new AtomicBoolean(false);
+        this.isClone = false;
     }
 
     @Override
-    protected void doReadInternal(ByteBuffer b) throws IOException {
+    protected void readInternal(ByteBuffer b) throws IOException {
```

Comment:
This one and `close` I duplicated outright and made `maybeReadChecksumFromFileInfo` static to make it work without inheritance. Didn't see a reasonable way of drying this up much.